bubbles/commands/periodic/rule_monitoring.py

```diff
@@ -0,0 +1,356 @@
+    """Get a message reflecting the edited rules."""
+    if len(changes["edited"]) == 0:
+        return None
+
+    def _edit_text(edit: RuleEdited) -> str:
+        index = edit["new_rule"]["index"]
+        heading = f"*## Edited Rule {index}*"
+        old = _format_rule(edit["old_rule"])
+        new = _format_rule(edit["new_rule"])
+        return f"{heading}\n\n*### Old*\n\n{old}\n\n*### New*\n\n{new}"
+
+    edited_rule_text = [_edit_text(edit) for edit in changes["edited"]]
+    return "\n\n".join(edited_rule_text)
+
+
+def _get_rule_added_message(changes: RuleChanges) -> Optional[str]:
+    """Get a message reflecting the added rules."""
+    if len(changes["added"]) == 0:
+        return None
+
+    added_rule_text = [
+        f"*## Added Rule {rule['index']}*\n\n{_format_rule(rule)}"
+        for rule in changes["added"]
+    ]
+    return "\n\n".join(added_rule_text)
+
+
+def _get_rule_removed_message(changes: RuleChanges) -> Optional[str]:
+    """Get a message reflecting the removed rules."""
+    if len(changes["removed"]) == 0:
+        return None
+
+    removed_rule_text = [
+        f"*## Removed Rule {rule['index']}*\n\n{_format_rule(rule)}"
+        for rule in changes["removed"]
+    ]
+    return "\n\n".join(removed_rule_text)
+
+
+def _get_rule_change_message(sub_name: str, changes: RuleChanges) -> Optional[str]:
+    """Get a message reflecting the rule changes for the given sub.
+
+    If there were no changes, `None` is returned.
+    """
+    edited_msg = _get_rule_edited_message(changes)
+    added_msg = _get_rule_added_message(changes)
+    removed_msg = _get_rule_removed_message(changes)
+
+    if edited_msg is None and added_msg is None and removed_msg is None:
+        # No changes
+        return None
+
+    msg = f"*# Rule changes in r/{sub_name}*"
+
+    if edited_msg:
+        msg += f"\n\n{edited_msg}"
+    if added_msg:
+        msg += f"\n\n{added_msg}"
+    if removed_msg:
+        msg += f"\n\n{removed_msg}"
+
+    return msg
+
+
+def _notify_mods(text: str) -> None:
+    """Send a message regarding the rule changes to the mods."""
+    app.client.chat_postMessage(
+        channel=RULE_MONITORING_CHANNEL_ID,
+        text=text,
+        link_names=1,
+        unfurl_links=False,
+        unfurl_media=False,
+        as_user=True,
+    )
+
+
+def rule_monitoring_callback():
+    """Check for rule changes for the next subreddit in the list.
+
+    If no subs are left to process, the subreddit list is updated again.
+    If new subs have been added, they will be processed immediately.
+    """
+    # Repopulate the subreddit queue if necessary
+    if len(subreddit_stack) == 0:
+        _initialize_subreddit_stack()
+
+    # If there are new subs, check all of their rules directly
+    if len(new_subreddits) > 0:
+        for sub_name in new_subreddits:
+            _initialize_rules(sub_name)
+            # Make sure we don't go over the API rate limit
+            sleep(1)
```

Comment:
if len(new_subreddits) ==0 or =0, what happens? could that be an issue? or this dealt with in line 331-332

Reply:
If `len(new_subreddits) == 0` then we don't need to do anything special, that means no new subs were added since the last time we checked.
So this is expected to be the case most of the time. Then we usually have `len(subreddit_stack) > 0` though, so we do not return in line 346 and instead process one of the subreddits on the stack in line 350 onwards.